### PR TITLE
 Customize coverage report by changing settings in the customize form

### LINKED ISSUE #155: Create the endpoint for the report page
HTML and PDF export

Since it's quite a lot of changes, perhaps let's split it into steps:

- [ ] Create standard report template and make it available in the app
- [ ] Create demo report endpoint to display report for demo sample (so we know it works
- [ ] Create report endpoint for normal samples
- [ ] Make report exportable to PDF

### PR DESCRIPTION
**Pull request by northwestwitch**
### This PR adds | fixes:
- Update report by changing settings in the filter form **- FINAL THING FOR THE REPORT! -** fix #155

Fixed customize report form to send data to report endpoint as json:

image

**How to prepare for test**:
- Deploy on cg-vm1

### How to test:
- Go to the demo report and modify it by changing the settings in the customise window (gene list, panel name, default threshold values)

### Expected outcome:
- [x] The report should update

### Review:
- [x] Code approved by HS
- [x] Tests executed by CR

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: src/chanjo2/templates/report.html [hunk 1]
```diff
@@ -16,7 +16,6 @@ <h4 class="panel-title">
 		</div>
 		<div id="view-filters" class="panel-collapse collapse">
 			<div class="panel-body">
-				<form method="POST" action="{{ url_for('report') }}" id="coverageForm")>
```
Comment (northwestwitch) [inline src/chanjo2/templates/report.html:19]: Here I decided to remove the form submit because the report endpoint accepts json data (form POST method is sending `application/x-www-form-urlencoded` data instead, which is not compatible with the existing endpoint).

Took me a while to figure out this because I'm still relatively new to FastAPI, but a good explanation is given in [here](https://stackoverflow.com/questions/71108731/how-to-include-json-and-file-data-together-in-fastapi-endpoint).

So basically I wrote a function that once you click "update" submits the form inputs as json, then collects the HTML code that returns from in the response and updates the html body with the new content.

File: src/chanjo2/templates/report.html [hunk 2]
```diff
@@ -301,6 +299,48 @@ <h3>Explanations</h3>
...
+			// Returns the list of HGNC genes present in the user query as a sring as a list of integers
+			function getFormGeneIDsAsList()
+			{
+				return document.getElementById("hgnc_gene_ids").value.replace(/\s/g, '').split(",").map(Number);
+			}
+
+			function submitAsJson()
+			{
+				event.preventDefault();
+				let formDict = {
```
Comment (northwestwitch) [inline src/chanjo2/templates/report.html:317]: This gets submitted in a POST request as JSON

File: src/chanjo2/templates/report.html [hunk 3]
```diff
@@ -301,6 +299,48 @@ <h3>Explanations</h3>
...
+				fetch('/report', {
+                        method: 'POST',
+                        headers: {
+                            'Content-Type': 'application/json'
+                        },
+                        body: JSON.stringify(formDict)
+                    })
+                    .then(resp => resp.text())
+                    .then(html => {
+                    	document.body.innerHTML = updatedReportBody(html);
```
Comment (northwestwitch) [inline src/chanjo2/templates/report.html:337]: The document body is replaced with the body of the HTML from the response

**Comment by northwestwitch** [comment]
Default demo panel - note the changes in panel name, default coverage level and the overall stats

image



Demo panel after submitting "the form":

image

File: src/chanjo2/models/pydantic_models.py [hunk 4]
```diff
@@ -225,6 +225,12 @@ class ReportQuery(BaseModel):
     samples: List[ReportQuerySample]
     case_display_name: Optional[str]
 
+    @validator("samples", pre=True)
+    def samples_validator(cls, value):
```
Comment (henrikstranneheim) [inline src/chanjo2/models/pydantic_models.py:229]: Is there a more descriptive name than`value` - Pydantic allows naming other than the generic value in validators.

**Review by henrikstranneheim — APPROVED** [review]
Well done 👍

File: src/chanjo2/templates/report.html (hunk 2 above)
Comment (northwestwitch) [reply src/chanjo2/templates/report.html:317]: Sure, I'll fix!

File: src/chanjo2/templates/report.html (hunk 3 above)
Comment (northwestwitch) [reply src/chanjo2/templates/report.html:337]: Fixed!
